src/MigrationTools.Clients.AzureDevops.ObjectModel/TfsExtensions.cs

```diff
@@ -83,38 +83,50 @@ public static string ToJson(this Field f)
             return JsonConvert.SerializeObject(expando, Formatting.Indented);
         }
 
-        public static void RefreshWorkItem(this WorkItemData context)
+        public static void RefreshWorkItem(this WorkItemData context, FieldCollection fieldsOfRevision = null)
         {
-            Log.Debug("TfsExtensions::RefreshWorkItem");
             var workItem = (WorkItem)context.internalObject;
-            context.Id = workItem.Id.ToString();
-            context.Type = workItem.Type.Name;
-            context.Title = workItem.Title;
-            context.Rev = workItem.Rev;
-            context.RevisedDate = workItem.RevisedDate;
-            context.Revision = workItem.Revision;
-            context.ProjectName = workItem?.Project?.Name;
-            context.Fields = workItem.Fields.AsDictionary();
-            context.Revisions = (from Revision x in workItem.Revisions
-                                 select new _EngineV1.DataContracts.RevisionItem()
-                                 {
-                                     Index = x.Index,
-                                     Number = Convert.ToInt32(x.Fields["System.Rev"].Value),
-                                     ChangedDate = Convert.ToDateTime(x.Fields["System.ChangedDate"].Value)
-                                 }).ToList();
+            context.ProjectName = workItem.Project?.Name;
+
+            // If fieldsOfRevision is provided we use this collection as we want to create a revised WorkItemData object
+            context.Fields = fieldsOfRevision != null ? fieldsOfRevision.AsDictionary() : workItem.Fields.AsDictionary();
+
+            // We only need to fill the revisions object if we create a WorkItemData object for the whole WorkItem and
+            // we sort it here by Number using a SortedDictionary
+            context.Revisions = fieldsOfRevision == null ? new SortedDictionary<int, RevisionItem>((from Revision x in workItem.Revisions
+                                                                                                    select new RevisionItem()
+                                                                                                    {
+                                                                                                        Index = x.Index,
+                                                                                                        Number = (int)x.Fields["System.Rev"].Value,
+                                                                                                        ChangedDate = (DateTime)x.Fields["System.ChangedDate"].Value,
+                                                                                                        Type = x.Fields["System.WorkItemType"].Value as string,
+                                                                                                        Fields = x.Fields
```

Comment:
This is the most relevant part and I forgot to mention it in my comment. This is the one and only place where fields of the revision are stored. In fact when we do a `GetWorkItem` with the revision overload it works but our ToWorkItemData overwrites the values again as `RefreshWorkItem` loads the values from the work item again not from the revision!